cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamLeaderManager.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.stream.service;
+
+import co.cask.cdap.api.data.stream.StreamSpecification;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.zookeeper.coordination.BalancedAssignmentStrategy;
+import co.cask.cdap.common.zookeeper.coordination.PartitionReplica;
+import co.cask.cdap.common.zookeeper.coordination.ResourceCoordinator;
+import co.cask.cdap.common.zookeeper.coordination.ResourceCoordinatorClient;
+import co.cask.cdap.common.zookeeper.coordination.ResourceHandler;
+import co.cask.cdap.common.zookeeper.coordination.ResourceModifier;
+import co.cask.cdap.common.zookeeper.coordination.ResourceRequirement;
+import com.google.common.base.Functions;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import org.apache.twill.api.ElectionHandler;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.internal.zookeeper.LeaderElection;
+import org.apache.twill.zookeeper.ZKClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Distributed implementation of the {@link StreamLeaderManager} that uses a {@link ResourceCoordinator} to
+ * allocate Streams equally amongst Stream handlers. One Stream handler will be the leader of a collection of streams.
+ * This ensures that no one stream handler gets overloaded with this task.
+ */
+public class DistributedStreamLeaderManager extends AbstractIdleService implements StreamLeaderManager {
```

Comment:
Well, the `StreamCoordinator` is to do stream handler/service coordination. Electing leader per stream is definitely falls into that category, right?